refactoring configs

hstojic opened the pull request:
This PR will close #373 where we chose to go with option 1a and 1b

Implemented
1. added a registry of models to enable config via dictionary again
2. improved tests
3. removed `create_optimizer`, now dealt with via model registry
4. notebook explaining configs
5. sorted out optimizers in gpflux (incompatible with ModelConfig)

hstojic committed:
removed configs from notebooks

hstojic committed:
registry added; tests implemented

hstojic committed:
fixed imports

hstojic committed:
notebook about configs

hstojic committed:
Merge branch 'develop' into hstojic/refactoring_configs

hstojic committed:
gfixing an import statement

hstojic committed:
black

hstojic committed:
modified optimizers in GPflux

hstojic committed:
removing warnings

apaleyes commented on docs/notebooks/failure_ego.pct.py:
```diff
@@ -117,26 +115,19 @@ def create_classification_model(data):
+    OBJECTIVE: GaussianProcessRegression(regression_model),
+    FAILURE: VariationalGaussianProcess(
+        classification_model,
+        BatchOptimizer(tf.optimizers.Adam(1e-3), max_iter=50),  
```
That looks like slight step back in terms of user experience. Now they need to know which optimizer class to use, which isn't obvious. E.g. I don't have a slight clue why we use `BatchOptimizer` here

hstojic replied:
here we need to explicitly pass the `BatchOptimizer` object because we are using non-default settings (`max_iter=50`), otherwise the wrapper tells you which `Optimizer` you need to use

I think there is no way around it, as soon as they need a non-default optimizer they need to know about it
what we could do is the following:
1. Explicitly educate users in one of the intro notebooks on difference between `Optimizer` and `BatchOptimizer` and when to use them
2. Eliminate it from the rest of the notebooks and simply rely on defaults set in the interface, notebooks should work fine with the defaults

apaleyes replied:
If there is a nice default, i'd go with option 2.

hstojic replied:
ah, these weren't meant to be options, I was proposing doing both things :)

apaleyes commented on docs/notebooks/model_config.pct.py:
```diff
@@ -0,0 +1,146 @@
+# %% [markdown]
+# ## Finding a minimum of the Branin function
+#
+# In this example, as in most other tutorial, we look to find the minimum value of the familiar two-dimensional Branin function over the hypercube $[0, 1]^2$.
```
```suggestion
# In this example, as in many other tutorials, we look to find the minimum value of the familiar two-dimensional Branin function over the hypercube $[0, 1]^2$.
```

apaleyes commented on docs/notebooks/model_config.pct.py:
```diff
@@ -0,0 +1,146 @@
+# %% [markdown]
+# # Building models with configuration dictionaries
+#
+# If you are an expert user of Trieste and some modelling library, GPflow for example, then building models via a configuration dictionary might be a useful alternative to working with model and optimizer interfaces. Here we provide an overview of how to use configuration dictionaries.
```
```suggestion
# If you are an expert user of Trieste and some modelling library, GPflow for example, then building models via a configuration dictionary might be a useful alternative to working with model and optimizer wrappers. Here we provide an overview of how to use configuration dictionaries.
```

apaleyes commented on docs/notebooks/model_config.pct.py:
```diff
@@ -0,0 +1,146 @@
+#
+# The Bayesian optimization procedure estimates the next best points to query by using a probabilistic model of the objective. We'll use Gaussian Process (GP) regression in this tutorial, as provided by GPflow.
+#
+# GPflow models cannot be used directly in our Bayesian optimization routines, only through a valid model interface. Trieste typically has a separate interface for each model. For instance, `GPR` model from GPflow has to be used with `GaussianProcessRegression` interface. These interfaces standardise outputs from all models, deal with preparation of the data and implement additional methods neede ...
```
That sounds slightly weird, as though each model class in GPflow has a corresponding class in Trieste. Maybe "model type"?

hstojic replied:
well, only SGPR and GPR share the wrapper GaussianProcessRegression, SVGP and VGP have their own wrapper...

apaleyes replied:
But at least it isn't designed "class per model", right? In theory many models can share one wrapper

hstojic replied:
that's true, I'll word it more carefully

apaleyes commented on docs/notebooks/model_config.pct.py:
```diff
@@ -0,0 +1,146 @@
+gpflow_model = build_model(initial_data)
+
+# %% [markdown]
+# Before using the model interface we might want to set the optimizer for it as well
+# and set some arguments that we want to pass to it. We need to use Trieste's optimizer
+# interfaces, here `Optimizer` would be the suitable interface. We'll optimize our model with GPflow's Scipy optimizer.
```
Same comment as above - even for simple use cases we seem to require knowledge of a suitable optimizer. Can this be simplified?

hstojic replied:
in simple cases we don't, only when we want to use non-default settings - perhaps I can spell that out here?

apaleyes replied:
That would be suitable place to explain this indeed

apaleyes commented on docs/notebooks/model_config.pct.py:
```diff
@@ -0,0 +1,146 @@
+# %% [markdown]
+# ## Using configuration dictionaries
+#
+# Instead of working direclty with model and optimization interfaces, after you know them sufficiently well, you can skip them by using configuration dictionary. It consists of a dictionary with same four arguments that can be passed to any model interface: `model`, `model_args`, `optimizer` and `optimizer_args`.
```
```suggestion
# Instead of working directly with model and optimization interfaces, after you know them sufficiently well, you can skip them by using configuration dictionary. It consists of a dictionary with same four arguments that can be passed to any model interface: `model`, `model_args`, `optimizer` and `optimizer_args`.
```

apaleyes commented on trieste/models/optimizer.py:
```diff
@@ -14,16 +14,15 @@
-register them using a :func:`create_optimizer`, as well as register their loss functions using
-a :func:`create_loss_function`.
+with models. Specific models can also sub-class these optimizers or implement their own, and should
+register register their loss functions using a :func:`create_loss_function`.
```
```suggestion
register their loss functions using a :func:`create_loss_function`.
```

apaleyes commented on trieste/models/gpflow/config.py:
```diff
@@ -14,35 +14,22 @@
+_SUPPORTED_MODELS: Dict[Type[Any], Tuple[Type[TrainableProbabilisticModel], Type[Optimizer]]] = {
+    GPR: (GaussianProcessRegression, Optimizer),
+    SGPR: (GaussianProcessRegression, Optimizer),
+    VGP: (VariationalGaussianProcess, BatchOptimizer),
```
As far as I can tell you can only know this correspondence either from here or from expert knowledge. Neither are reasonable to expect from non-power users. I think that makes our default class-based method quite hard to use. Can we consider maybe default optimizers for models, or using registry for models? Or am I overthinking this?

hstojic replied:
does looking up a wrapper to figure out what arguments are needed, say `GaussianProcessRegression`, constitute an expert knowledge?

each wrapper does use default optimizers, perhaps we are just not doing ourselves a favour by not using them in notebooks?

apaleyes replied:
Somewhat... let's use defaults in notebooks as much as we can, so that readers don't start asking "why" or "is that important" a lot.

apaleyes reviewed (commented):
If you don't count the notebook, this PR removes more lines than it adds. Great refactor!

However I do have two main comments:
1. Notebook not very convincing, as it doesn't explain why would anyone want to use configs or a registry. Can we at least simulate a bit more exciting use case?
2. Most of notebooks used to not require expert knowledge of which optimizer wrapper shall be used. Now they do. I feel that's a step back in user experience. Can we avoid that?

hstojic commented:
> If you don't count the notebook, this PR removes more lines than it adds. Great refactor!
> 
> However I do have two main comments:
> 
> 1. Notebook not very convincing, as it doesn't explain why would anyone want to use configs or a registry. Can we at least simulate a bit more exciting use case?
> 2. Most of notebooks used to not require expert knowledge of which optimizer wrapper shall be used. Now they do. I feel that's a step back in user experience. Can we avoid that?

I completely agree about 1), but not sure we can come up with a better use case for configs. I can try setting up a use case for running experiments that we discussed, perhaps that would end up being a better one? 

Regarding 2), we could either use default optimizers in notebooks and not specify them, or go back to using dictionaries which would imply that setting up models with a dictionary is for beginners, not experts? I would be up for the former option, I think even beginners need some exposure to model wrappers so that they know where to look up the info about more advanced usage

apaleyes commented:
> I completely agree about 1), but not sure we can come up with a better use case for configs. I can try setting up a use case for running experiments that we discussed, perhaps that would end up being a better one?
> 
> Regarding 2), we could either use default optimizers in notebooks and not specify them, or go back to using dictionaries which would imply that setting up models with a dictionary is for beginners, not experts? I would be up for the former option, I think even beginners need some exposure to model wrappers so that they know where to look up the info about more advanced usage

Just mentioning it in the notebook would be fine for now. And yeah, let's go with defaults where we can

hstojic committed:
addressing some of the comments

hstojic committed:
merged in develop

hstojic committed:
fixes

hstojic committed:
Merge branch 'develop' into hstojic/refactoring_configs

hstojic committed:
Andrei's comments addressed, notebook added a use case; formatting

hstojic committed:
fixed a missing import

hstojic commented:
Addressed all the comments, here are the highlights: 
- removed custom optimizer from most notebooks (not really needed) 
- added a more detailed explanation on role of wrappers in an intro notebook
- added another use case to the model_config notebook

hstojic committed:
removing isort related changes

apaleyes commented on docs/notebooks/active_learning.pct.py:
```diff
@@ -45,7 +45,7 @@
 # ## Surrogate model
 #
-# Just like in sequential optimization, we fit a surrogate Gaussian process model to the initial data.
+# Just like in sequential optimization, we fit a surrogate Gaussian process model as implemented in GPflow to the initial data. As usual, the GPflow models cannot be used directly in our Bayesian optimization routines, only through a valid model wrapper. Below we construct a `GPR` model from GPflow and pass it to the appropriate `GaussianProcessRegression` wrapper.
```
```suggestion
# Just like in sequential optimization, we fit a surrogate Gaussian process model as implemented in GPflow to the initial data. The GPflow models cannot be used directly in our Bayesian optimization routines, only through a valid model wrapper. Below we construct a `GPR` model from GPflow and pass it to the appropriate `GaussianProcessRegression` wrapper.
```

it's not usual, they just can't

uri-granta replied:
I expect the purpose of the "as usual" is to highlight that this bit of information is not new, but one the reader should have previously encountered when reading the tutorials. Though I agree with your later comment that we don't need to be so verbose each time we mention this.

apaleyes commented on docs/notebooks/batch_optimization.pct.py:
```diff
@@ -35,13 +35,13 @@
 # ## Surrogate model
-# Just like in purely sequential optimization, we fit a surrogate Gaussian process model to the initial data.
+#
+# Just like in purely sequential optimization, we fit a surrogate Gaussian process model to the initial data. As usual, the GPflow models cannot be used directly in our Bayesian optimization routines, only through a valid model wrapper. Below we construct a `GPR` model from GPflow and pass it to the appropriate `GaussianProcessRegression` wrapper.
```
Is it actually worth adding this bit in many notebooks? We can be fairly detailed in one notebook, and refer to it thereafter.

apaleyes commented on docs/notebooks/model_config.pct.py:
```diff
@@ -110,19 +113,79 @@ def build_model(data):
+# %% [markdown]
+# ## Using configuration dictionaries for setting up experiments
+#
+# Another use case is in setting up experiments, where it becomes easier to benchmark Bayesian optimization algorithms. The advantage is that we can easily change the models and set up any argument for them from one condition to another - only object with the experiment specification needs to change (`experiment_conditions` below), while the rest of the code for executing experiments can stay the ...
```
Second sentence is a bit hard to read, can you simplify it? Also what are these "conditions"?

uri-granta replied:
Agree. Also, replace - with : or —.

apaleyes commented on docs/notebooks/model_config.pct.py:
```diff
@@ -110,19 +113,79 @@ def build_model(data):
+}
+
+# here we specify our experiments
+experiment_conditions = [
+    {"model_args": {"num_kernel_samples": 500}},
+    {"model_args": {"num_kernel_samples": 1000}},
+    {"model": build_dgp_model(initial_data)},
+]
```
To make sure this doesn't affect CI you may want to leave only one, or none at all, for quick test purposes.

uri-granta replied:
Agree. Feel free to ask me if you're unsure how to set up the quicktest overrides.

hstojic replied:
gave it a try, @uri-granta how can I check whether quickrun is working properly?

apaleyes replied:
Maybe you could try running quicktest.py locally and see what notebook it creates?

hstojic replied:
its not working, probably regular expression in the quickrun.py is not that flexible, I suggest we leave it as is, its two BO runs now with 2 iterations only, so should be quick enough

apaleyes reviewed (commented):
Looks good! And experimentation example in the notebook illustrates utility of configs well.

uri-granta commented on docs/notebooks/deep_gaussian_processes.pct.py:
```diff
@@ -69,14 +69,16 @@
+# Since DGPs can be hard to build, Trieste provides some basic architectures: here we use the `build_vanilla_deep_gp` function which returns a GPflux model of `DeepGP` class. As with other models (e.g. GPflow), we cannot use it directly in Bayesian optimization routines, we need to pass it through an appropriate wrapper, `DeepGaussianProcess` wrapper in this case.
 #
-# Since DGPs can be hard to build, Trieste provides some basic architectures: here we use the `build_vanilla_deep_gp` method.
+# Few other useful notes regarding building a DGP model. The DGP model requires us to specify the number of inducing points, as we don't have the true posterior. To train the model we have to use a stochastic optimizer; Adam is used by default, but we can use other stochastic optimizers from TensorFlow. GPflux allows us to use the Keras `fit` method, which makes optimizing a lot easier - this met ...
```
(we should probably be more consistent regarding whether we say trieste or Trieste)

hstojic replied:
do we use trieste or Trieste everywhere?

hstojic replied:
went with Trieste and changed few more notebooks that used trieste

uri-granta commented on docs/notebooks/failure_ego.pct.py:
```diff
@@ -105,38 +104,29 @@ def create_classification_model(data):
 # We now specify how Trieste will use our GPflow models within the BO loop.
 #
-# For our `GPR` model, we will use a standard L-BFGS optimizer from Scipy, whereas we will optimze our `VGP` model using alternate Adam steps (to optimize kernel parameter) and NatGrad steps (to optimize variational parameters).
+# For our `VGP` model we will use a non-default optimization: alternate Adam steps (to optimize kernel parameter) and NatGrad steps (to optimize variational parameters). For this we need to use the `BatchOptimizer` wrapper and set the `use_natgrads` model argument to `True` in our `VariationalGaussianProcess` model wrapper.
```
so is this notebook currently only working because we use a custom acquisition function that's not compiled (which means it doesn't hit the bug I'm trying to resolve)?

hstojic replied:
probably

uri-granta commented on docs/notebooks/model_config.pct.py:
```diff
@@ -0,0 +1,209 @@
+#
+# The Bayesian optimization procedure estimates the next best points to query by using a probabilistic model of the objective. We'll use Gaussian Process (GP) regression in this tutorial, as provided by GPflow.
+#
+# The GPflow models cannot be used directly in our Bayesian optimization routines, only through a valid model wrapper. Trieste has wrappers that support several popular model. For instance, `GPR` and `SGPR` models from GPflow have to be used with `GaussianProcessRegression` wrapper. These wrappers standardise outputs from all models, deal with preparation of the data and implement additional meth ...
```
```suggestion
# The GPflow models cannot be used directly in our Bayesian optimization routines, only through a valid model wrapper. Trieste has wrappers that support several popular models. For instance, `GPR` and `SGPR` models from GPflow have to be used with the `GaussianProcessRegression` wrapper. These wrappers standardise outputs from all models, deal with preparation of the data and implement additional methods needed for Bayesian optimization.
```

uri-granta commented on docs/notebooks/model_config.pct.py:
```diff
@@ -0,0 +1,209 @@
+gpflow_model = build_model(initial_data)
+
+# %% [markdown]
+# Usually constructing a GPflow model would be enough, as it is the only required argument for the model wrappers. Wrappers have other arguments - an `optimizer` argument as a rule and potentially some additional model arguments (for example, `num_kernel_samples` in `GaussianProcessRegression`). These arguments are set to sensible defaults and hence typically we can simplify the model building. 
```
(or if you're really wedded to using dashes then use an em-dash — )

```suggestion
# Usually constructing a GPflow model would be enough, as it is the only required argument for the model wrappers. Wrappers have other arguments: an `optimizer` argument as a rule and potentially some additional model arguments (for example, `num_kernel_samples` in `GaussianProcessRegression`). These arguments are set to sensible defaults and hence typically we can simplify the model building. 
```

uri-granta commented on docs/notebooks/model_config.pct.py:
```diff
@@ -0,0 +1,209 @@
+
+results = []
+for exp in experiment_conditions:
+    model_config = deepcopy(basic_config)
```
Should we provide a `combine_dict_configs` function that merges two dict configs? This would simplify this code and also handle properly merging nested configs (e.g. if one of the experiments tried to override maxiter).

```
def merge_dict_configs(config1: ModelDictConfig, config2: ModelDictConfig) -> ModelDictConfig:
   ...
```

hstojic replied:
this is really an illustration only, I don't want to say this is the best way to do it - I'll add a comment reflecting this

uri-granta commented on trieste/models/config.py:
```diff
@@ -12,100 +12,181 @@
 
-
-ModelSpec = Union[ModelConfig, TrainableProbabilisticModel]
+ModelSpec = Union[Dict[str, Any], ModelConfig, TrainableProbabilisticModel]
```
Might be useful to also define an alias

```
ModelDictConfig = Dict[str, Any]
```

uri-granta commented on docs/notebooks/model_config.pct.py:
```diff
@@ -0,0 +1,209 @@
+# %% [markdown]
+# ## Registry of supported models
+#
+# Configuration dictionaries are made possible with the `ModelRegistry` that contains mapping between each model (e.g. GPflow or GPflux) and corresponding model wrapper and optimizer wrapper. All models that Trieste currently supports are registered there.
```
```suggestion
# Configuration dictionaries are made possible with the `ModelRegistry` that contains mapping between each model (e.g. GPflow or GPflux) and the corresponding model wrapper and optimizer wrapper. All models that Trieste currently supports are registered there.
```

uri-granta commented on docs/notebooks/model_config.pct.py:
```diff
@@ -0,0 +1,209 @@
+#
+# Configuration dictionaries are made possible with the `ModelRegistry` that contains mapping between each model (e.g. GPflow or GPflux) and corresponding model wrapper and optimizer wrapper. All models that Trieste currently supports are registered there.
+#
+# You can add new models to the registry, in case you have custom models with which you wish to use the configuration dictionaries. Let see an example of this. We will register the `GPMC` model from GPflow that is currently not supported. You would likely need to create a new model wrapper and perhaps a new optimizer wrapper as well, but just for the sake of an example we will borrow here existin ...
```
```suggestion
# You can add new models to the registry, in case you have custom models with which you wish to use the configuration dictionaries. Let's see an example of this. We will register the `GPMC` model from GPflow that is currently not supported. You would likely need to create a new model wrapper and perhaps a new optimizer wrapper as well, but just for the sake of an example we will borrow here existing wrappers, `GaussianProcessRegression` and `Optimizer`.
```

uri-granta commented on docs/notebooks/model_config.pct.py:
```diff
@@ -0,0 +1,209 @@
+ModelRegistry.register_model(gpflow.models.GPMC, GaussianProcessRegression, Optimizer)
+
+# check if it has been registered
+print(gpflow.models.GPMC in ModelRegistry.get_registered_models())
```
Is the print necessary?

```suggestion
gpflow.models.GPMC in ModelRegistry.get_registered_models()
```

hstojic replied:
it is

uri-granta commented on trieste/models/gpflow/config.py:
```diff
@@ -14,35 +14,22 @@
+    VGP: (VariationalGaussianProcess, BatchOptimizer),
+    SVGP: (SparseVariational, BatchOptimizer),
+}
+for model_type, interface_optimizer in _SUPPORTED_MODELS.items():
```
You can do this:

```suggestion
for model_type, (interface, optimizer) in _SUPPORTED_MODELS.items():
```

uri-granta commented on trieste/models/gpflux/config.py:
```diff
@@ -14,35 +14,19 @@
+_SUPPORTED_MODELS: Dict[Type[Any], Tuple[Type[TrainableProbabilisticModel], Type[Optimizer]]] = {
+    DeepGP: (DeepGaussianProcess, Optimizer),
+}
+for model_type, interface_optimizer in _SUPPORTED_MODELS.items():
```
```suggestion
for model_type, (interface, optimizer) in _SUPPORTED_MODELS.items():
```

uri-granta commented on trieste/models/config.py:
```diff
@@ -12,100 +12,181 @@
+        """
+        if model_type in cls._REGISTRY.keys():
+            warn(
+                f"Model {model_type} has already been registered, " f"you have now overwritten it. "
```
```suggestion
                f"Model {model_type} has already been registered, you have now overwritten it."
```

uri-granta commented on tests/util/models/gpflow/models.py:
```diff
@@ -116,6 +116,10 @@ def get_kernel(self) -> tfp.math.psd_kernels.PositiveSemidefiniteKernel:
         return self.kernel
 
 
+class GPRcopy(GPR):
```
Is this only used for testing the registry? If so it probably belongs there rather than here (as it's not a generally useful class).

uri-granta commented on tests/unit/models/test_config.py:
```diff
@@ -14,18 +14,61 @@
+
+def test_model_registry_register_model_warning() -> None:
+
+    with pytest.warns(UserWarning) as record:
```
It's generally bad to have tests that depend on other tests being run first. You can fix it here (and below) by explicitly (re)registering first. (It would be even better to ensure the global state is put back at the end of each of the tests, but that's harder without adding an unregister_model which probably isn't worth it.)

uri-granta commented on tests/unit/models/gpflow/test_config.py:
```diff
@@ -14,72 +14,39 @@
-    assert isinstance(model_config.optimizer, expected_optimizer)
+    assert model_type in ModelRegistry.get_registered_models()
+    assert ModelRegistry.get_interface(model_type) == interface
+    assert ModelRegistry.get_optimizer(model_type) == optimizer
```
Any reason not to (also) build a ModelConfig like before and check the types of the actual constructed optimizer?

hstojic replied:
huh, not sure whether this is the right place anymore for that - wrappers providing defaults for the optimizers, so this should probably be done in `test_models.py`, not really related to configs
(before it was related to model config's default for the optimizer, but we don't have that anymore, that's something that wrappers should deal with)

uri-granta commented on tests/unit/models/gpflux/test_config.py:
```diff
@@ -14,66 +14,29 @@
+    model_type, interface, optimizer = supported_models
+
+    assert model_type in ModelRegistry.get_registered_models()
+    assert ModelRegistry.get_interface(model_type) == interface
```
(same comment as above)

uri-granta reviewed (approved):
Overall looks good. Various comments (feel free to ask for clarifications if required).

hstojic committed:
addressed most of the review comments

hstojic committed:
addressed comments; VGP corrections

hstojic committed:
Merge branch 'develop' into hstojic/refactoring_configs

hstojic committed:
merged develop

hstojic committed:
quickrun for the notebook
